mmv1/products/vertexai/api.yaml

```diff
@@ -102,4 +102,173 @@ objects:
+      guides:
+        'Official Documentation':
+          'https://cloud.google.com/vertex-ai/docs'
+      api: 'https://cloud.google.com/vertex-ai/docs/reference/rest/v1beta1/projects.locations.featurestores'
+    async: !ruby/object:Api::OpAsync
+      operation: !ruby/object:Api::OpAsync::Operation
+        path: 'name'
+        base_url: '{{op_id}}'
+        wait_ms: 1000
+      result: !ruby/object:Api::OpAsync::Result
+        path: 'response'
+        resource_inside_response: true
+      status: !ruby/object:Api::OpAsync::Status
+        path: 'done'
+        complete: True
+        allowed:
+          - True
+          - False
+      error: !ruby/object:Api::OpAsync::Error
+        path: 'error'
+        message: 'message'
+    description: |-
+      A collection of DataItems and Annotations on them.
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: region
+        description: The region of the dataset. eg us-central1
+        url_param_only: true
+        input: true
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: The name of the Featurestore. This value may be up to 60 characters, and valid characters are [a-z0-9_]. The first character cannot be a number.
+        input: true
+        url_param_only: true
+        pattern: projects/{{project}}/locations/{{region}}/featurestores/{{name}}
+      - !ruby/object:Api::Type::String
+        name: 'etag'
+        description: Used to perform consistent read-modify-write updates. 
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'createTime'
+        output: true
+        description: |
+          The timestamp of when the featurestore was created in RFC3339 UTC "Zulu" format, with nanosecond resolution and up to nine fractional digits.
+      - !ruby/object:Api::Type::String
+        name: 'updateTime'
+        output: true
+        description: |
+          The timestamp of when the featurestore was last updated in RFC3339 UTC "Zulu" format, with nanosecond resolution and up to nine fractional digits.
+      - !ruby/object:Api::Type::KeyValuePairs
+        name: 'labels'
+        description: |
+          A set of key/value label pairs to assign to this Workflow.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'onlineServingConfig'
+        description: |
+           Config for online serving resources.
+        properties:
+          - !ruby/object:Api::Type::Integer
+            name: 'fixedNodeCount'
+            required: true
+            description: |
+              The number of nodes for each cluster. The number of nodes will not scale automatically but can be scaled manually by providing different values when updating.
+# Vertex AI Featurestores
+  - !ruby/object:Api::Resource
+    name: FeaturestoreEntitytype
+    base_url: 'projects/{{project}}/locations/{{region}}/featurestores/{{featurestore}}/entityTypes/{{name}}'
```

Comment:
If I convert this to `'{{featurestore}}/entityTypes/{{name}}'`, it won't work thanks to has_project variable in resource.erb

What I want to do is:
- Remove project and region from schema. 
- A custom pre_create, pre_update and pre_delete snippet with the following code:

```
if v, ok := d.GetOk("featurestore"); ok {
	re := regexp.MustCompile("projects/([a-zA-Z0-9-]*)/(?:locations|regions)/([a-zA-Z0-9-]*)")
	switch {
	case re.MatchString(selfLink):
		if res := re.FindStringSubmatch(selfLink); len(res) == 3 && res[1] != "" {
			project = res[1]
			region = res[2]
		}
	}
}
```

Reply:
Hi @upodroid,
Looking more into this, I think the data catalog entryGroup + entry relationship is similar to this scenario. Take a look: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/products/datacatalog/api.yaml#L80

We could similarly separate out the base_url and create_url and add a `entityTypeId` url param. And then make `name` an output field and set the self_link to that. Looks like doing so makes import functionality tricky, so there is a custom import override as well: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/products/datacatalog/terraform.yaml#L79

Reply:
That won't work. DatacatalogEntry returns the Entry instead of an Operation. I tried that earlier

```
REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  16⬇  9✔  3✎  1⚑  $    make test
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
go generate  ./...
go test  -timeout=30s $(go list ./...)
# github.com/hashicorp/terraform-provider-google-beta/google-beta
google-beta/resource_vertex_ai_featurestore_entitytype.go:172:45: not enough arguments in call to vertexAIOperationWaitTimeWithResponse
        have (*Config, map[string]interface {}, *map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, *map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:298:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:336:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
note: module requires Go 1.16
# github.com/hashicorp/terraform-provider-google-beta/google-beta [github.com/hashicorp/terraform-provider-google-beta/google-beta.test]
google-beta/resource_vertex_ai_featurestore_entitytype.go:172:45: not enough arguments in call to vertexAIOperationWaitTimeWithResponse
        have (*Config, map[string]interface {}, *map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, *map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:298:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:336:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
note: module requires Go 1.16
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta [build failed]
make: *** [test] Error 2
```

One of the reasons I don't like the encoder is the potential issues with multi regional terraform configs. 

```
provider "google-beta" {
 region = "europe-north1"
}
resource "google_vertex_ai_featurestore" "featurestore" {
  provider = google-beta
  name     = "terraform%{random_suffix}"
  region   = "us-central1"
}

resource "google_vertex_ai_featurestore_entitytype" "featurestore" {
  provider = google-beta
  name     = "terraform%{random_suffix}"
  featurestore = google_vertex_ai_featurestore.featurestore.name
}
```

Reply:
nvm, my last commit fixed this issue.